go/httpdispatcher/proxy.go

```diff
@@ -150,42 +151,73 @@ func (p *Proxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 
 		_, PortalURL := p.detectPortalURL(r)
 
-		if (passThrough.checkDetectionMechanisms(ctx, fqdn.String()) || passThrough.URIException.MatchString(r.RequestURI)) && passThrough.SecureRedirect {
-			PortalURL.Scheme = "http"
-		}
 		log.LoggerWContext(ctx).Debug(fmt.Sprintln(host, "Redirect to the portal"))
-		PortalURL.RawQuery = "destination_url=" + r.Header.Get("X-Forwarded-Proto") + "://" + host + r.RequestURI
+
+		destURL, _ := url.Parse(r.URL.String())
+
+		destURL.Scheme = r.Header.Get("X-Forwarded-Proto")
+		if r.Referer() != "" {
+			refererHost, err := url.Parse(r.Referer())
+			if err != nil {
+				destURL.Host = host
+			} else {
+				destURL.Host = refererHost.Hostname()
+			}
+		} else {
+			destURL.Host = host
+		}
 
 		if parking {
 			PortalURL.Path = ""
 			PortalURL.RawQuery = ""
 		}
 
+		// Detect wispr user agent but exclude istuff
```

Comment:
lets call them iOS devices or "Apple stuff" if it applies to all their devices

istuff looks like stuff but you added an i at the beginning because you forgot you were already in insert mode